Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift

```diff
@@ -44,153 +45,90 @@ public enum ErrorCodes: Int {
-        additionalDataSize: UInt,
-        plainText: UnsafePointer<UInt8>,
-        plainTextSize: UInt,
-        cipherText: UnsafeMutablePointer<UInt8>
-    ) -> ErrorCodes {
+        key: SpanConstUInt8, iv: SpanConstUInt8, ad: SpanConstUInt8, message: SpanConstUInt8,
+        desiredTagLengthInBytes: Int
+    ) -> AesGcmRV {
+        var rv = AesGcmRV()
         do {
-            if keySize > Int.max
-                || ivSize > Int.max
-                || plainTextSize > Int.max
-                || additionalDataSize > Int.max
-            {
-                return .tooBigArguments
-            }
-            if ivSize == 0 {
-                return .invalidArgument
-            }
-            let nonce = try AES.GCM.Nonce(
-                data: UnsafeBufferPointer(start: iv, count: Int(ivSize)))
-            var message = UnsafeBufferPointer(start: plainText, count: Int(plainTextSize))
-            if plainTextSize == 0 {
-                Utils.zeroArray.withUnsafeBufferPointer { ptr in
-                    message = ptr
-                }
+            if iv.size() == 0 {
+                rv.errorCode = .invalidArgument
+                return rv
             }
-            let symKey = SymmetricKey(
-                data: UnsafeBufferPointer(start: key, count: Int(keySize)))
             let sealedBox: AES.GCM.SealedBox
-            if additionalDataSize > 0 {
+            if ad.size() > 0 {
                 sealedBox = try AES.GCM.seal(
-                    message,
-                    using: symKey,
-                    nonce: nonce,
-                    authenticating: UnsafeBufferPointer(
-                        start: additionalData, count: Int(additionalDataSize))
-                )
+                    Data.fromSpan(spanNoCopy: message),
+                    using: SymmetricKey(data: key),
+                    nonce: AES.GCM.Nonce(data: Data.fromSpan(spanNoCopy: iv)),
+                    authenticating: Data.fromSpan(spanNoCopy: ad))
             } else {
                 sealedBox = try AES.GCM.seal(
-                    message,
-                    using: symKey,
-                    nonce: nonce
+                    Data.fromSpan(spanNoCopy: message),
+                    using: SymmetricKey(data: key),
+                    nonce: AES.GCM.Nonce(data: Data.fromSpan(spanNoCopy: iv))
                 )
             }
-            sealedBox.ciphertext.copyBytes(to: cipherText, count: Int(plainTextSize))
-            sealedBox.tag.copyBytes(
-                to: cipherText + Int(plainTextSize), count: sealedBox.tag.count)
+            var ciphertext = VectorUInt8(sealedBox.ciphertext.count + desiredTagLengthInBytes)
+            var index = 0
+            sealedBox.ciphertext.forEach { val in
+                ciphertext[index] = val
+                index += 1
+            }
+            sealedBox.tag[..<desiredTagLengthInBytes].forEach { val in
+                ciphertext[index] = val
+                index += 1
+            }
```

Comment:
Copying these byte-by-byte into `ciphertext` is sad, especially across two loops. I'll just mention here that it'd be better if CryptoKit's SealBox could just serialize this for us.

Reply:
This is probably another case where we want to put an API shim inside UnsafeWrappers.swift.